### ISSUE TITLE
Deploy AzureRm App Service VSTS task fails after upgrading to Microsoft.Azure.DocumentDb nuget 1.11.1 from 1.9.5

### ISSUE THREAD

**Issue by Jacob-McKay** [Azure/azure-documentdb-dotnet#191]
I updated my docdb sdk so I could make use of async document queries.  The first hurdle was needing to change all my web apps and iis express instances to run in 64 bit.

However, when I went to deploy my webapp after the first successful deployment, I get file lock issues during deployment.  Here is the output of my deploy task:

2017-01-13T20:27:32.4280842Z ##[section]Finishing: Download Artifacts
2017-01-13T20:27:32.4320845Z ##[section]Starting: Deploy AzureRM App Service: CONFIDENTIAL
2017-01-13T20:27:32.4800852Z ==============================================================================
2017-01-13T20:27:32.4800852Z Task         : Azure App Service Deploy
2017-01-13T20:27:32.4800852Z Description  : Update Azure App Service using Web Deploy / Kudu REST APIs
2017-01-13T20:27:32.4810830Z Version      : 2.1.8
2017-01-13T20:27:32.4810830Z Author       : Microsoft Corporation
2017-01-13T20:27:32.4810830Z Help         : [More Information](https://aka.ms/azurermwebdeployreadme)
2017-01-13T20:27:32.4810830Z ==============================================================================
2017-01-13T20:27:33.4837534Z CONFIDENTIAL exists true
2017-01-13T20:27:36.4226911Z Got connection details for Azure App Service:'CONFIDENTIAL'
2017-01-13T20:27:36.4576921Z Running command: "C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:getParameters -source:package="C:\a\r1\a\CONFIDENTIAL.zip" > "C:\a\r1\a\parameter.xml"
2017-01-13T20:27:37.7526618Z Running command: @echo off 
2017-01-13T20:27:37.7526618Z "C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe"  -verb:sync -source:package="C:\a\r1\a\CONFIDENTIAL.zip" -dest:auto,ComputerName='https://CONFIDENTIAL.net:443/msdeploy.axd?site=CONFIDENTIAL',UserName='********',Password='********',AuthType='Basic' -setParam:name='IIS Web Application Name',value='CONFIDENTIAL' -enableRule:DoNotDeleteRule -userAgent:VSTS_CONFIDENTIAL_release_11_1272_3643_1 2>error.txt
2017-01-13T20:27:37.7536620Z if %errorlevel% neq 0 exit /b %errorlevel%
2017-01-13T20:27:37.7556609Z [command]C:\Windows\system32\cmd.exe /C C:\a\r1\a\msDeployCommand.bat
2017-01-13T20:27:38.4146621Z Info: Using ID 'CONFIDENTIAL' for connections to the remote server.
2017-01-13T20:27:41.0598530Z Info: Adding ACL's for path (CONFIDENTIAL)
2017-01-13T20:27:41.0598530Z Info: Adding ACL's for path (CONFIDENTIAL)
2017-01-13T20:27:41.0628531Z Info: Using ID 'CONFIDENTIAL' for connections to the remote server.
2017-01-13T20:27:46.1265841Z Info: Updating file (CONFIDENTIAL).
2017-01-13T20:27:48.1225406Z ##[error]Failed to deploy App Service.
2017-01-13T20:27:48.1255405Z ##[error]Error Code: ERROR_FILE_IN_USE
2017-01-13T20:27:48.1255405Z More Information: Web Deploy cannot modify the file 'DocumentDB.Spatial.Sql.dll' on the destination because it is locked by an external process.  In order to allow the publish operation to succeed, you may need to either restart your application to release the lock, or use the AppOffline rule handler for .Net applications on your next publish attempt.  Learn more at: http://go.microsoft.com/fwlink/?LinkId=221672#ERROR_FILE_IN_USE.
2017-01-13T20:27:48.1265401Z Error count: 1.
2017-01-13T20:27:48.1265401Z 
2017-01-13T20:27:49.4330113Z Successfully updated deployment History at https://CONFIDENTIAL/deployments/12721484339268116
2017-01-13T20:27:49.4330113Z ##[error]Error: Error: C:\Windows\system32\cmd.exe failed with return code: 4294967295
2017-01-13T20:27:49.4460098Z ##[section]Finishing: Deploy AzureRM App Service: CONFIDENTIAL
2017-01-13T20:27:49.4490179Z ##[section]Finishing: Release
 
I have tried using the 'take app offline' option in this deploy task, still no avail.  Any suggestions?

**Comment by rnagpal** [comment]
@Jacob-McKay 
1. Though moving to 64 bit process will give you best performance but yes that was a regression in 1.11.0 release and we will be releasing a fix early next week to allow 32 bit apps to run as they were until 1.10.0 release.
2. File lock issue is related to ASP.NET as I also faced similar issues. See https://github.com/aspnet/IISIntegration/issues/238#issuecomment-270210277 for more details. I haven't tried it yet.

**Comment by Jacob-McKay** [comment]
#2 explains why the take app offline option in VSTS msdeloy wasn't working.  Thanks @rnagpal !

**Comment by rnagpal** [comment]
@Jacob-McKay We have released .NET SDK 1.11.2 that supports 32-bit host process as before but we recommend using 64-bit host processing for improved performance.

**Comment by Jacob-McKay** [comment]
Hmm the DocumentDB.Spatial.Sql.dll file lock issue wasn't a problem in our releases until I upgraded the sdk.  Previously we were not using the 'take app offline' option, so whether or not 'take app offline' has a case sensitivity issue, I would still look to the sdk as the source of holding onto DocumentDB.Spatial.Sql.dll and failing our releases.

I can go 64 bit no problem, it's the file locking that is a more serious issue.  I will try your latest release when I have time to see if I it still breaks my releases.

**Comment by rnagpal** [comment]
@Jacob-McKay I believe it's related to the ASP.NET bits installed on your system. Try downgrading it to 1.9.5(that you had earlier) and see it you still get this file lock issue which would confirm ASP.NET issue. I'm not too much familiar with this particular feature of ASP.NET. 
@vijayrkn Can you please help unblock Jacob?
DocumentDB.Spatial.Sql.dll is one of the native dlls that we package in our Nuget package and binplace it along with our main Client dll as it's needed at runtime in some of the scenarios.

**Comment by vijayrkn** [comment]
From the above log, the msdeploy call looks good. 
```"C:\Program Files\IIS\Microsoft Web Deploy V3\msdeploy.exe" -verb:sync -source:package="C:\a\r1\a\CONFIDENTIAL.zip" -dest:auto,ComputerName='https://CONFIDENTIAL.net:443/msdeploy.axd?site=CONFIDENTIAL',UserName='********',Password='********',AuthType='Basic' -setParam:name='IIS Web Application Name',value='CONFIDENTIAL' -enableRule:DoNotDeleteRule -userAgent:VSTS_CONFIDENTIAL_release_11_1272_3643_1 2>error.txt```

so, the actually looks like the file is being locked by app. I am from the publish tooling team. Maybe Someone from the runtime side may be able to help.

**Comment by rnagpal** [comment]
@Jacob-McKay Let me know if you were able to resolve the file lock issue. There is nothing in the SDK that I'm aware of that locks this file. As I mentioned earlier, use the same project but instead use 1.9.5 to see if you still repro. If yes, this is most likely your app specific. Are you able to repro this consistently? If yes, it would help if you send the solution and I look into it.

**Comment by SteMcLellan** [comment]
@rnagpal I am experiencing the same issues as @Jacob-McKay.

We use `msdeploy.exe` to update our Azure App Service. When uploading we consistently receive the following messages:

> Info: Updating file (frabby-iso/comm\bin\DocumentDB.Spatial.Sql.dll).
> Error Code: ERROR_FILE_IN_USE
> More Information: Web Deploy cannot modify the file 'DocumentDB.Spatial.Sql.dll' on the destination because it is locked by an external process.  In order to allow the publish operation to succeed, you may need to either restart your application to release the lock, or use the AppOffline rule handler for .Net applications on your next publish attempt. Learn more at: http://go.microsoft.com/fwlink/?LinkId=221672#ERROR_FILE_IN_USE.

Our `msdeploy.exe` command is using the following parameters:

> -verb:sync -source:contentPath='C:\tmp\comm' -dest:contentPath='frabby-iso/comm',ComputerName='https://waws-prod-blu-013.publish.azurewebsites.windows.net/msdeploy.axd?site=frabby-iso',UserName=xxx,Password='xxx',AuthType='Basic' -enablerule:DoNotDeleteRule -enablerule:AppOffline -useCheckSum

Please notice that the error message states we should be using the AppOffline rule, and we are including it in our web deploy request. Occasionally the deployment will go through w/o error, but most of the time it fails. I can verify that after the site is up and running, that `DocumentDB.Spatial.Sql.dll` is locked by some process by navigating to that file through the KUDU Process Explorer for my Azure Website. 

The app that we are trying to deploy is too large to share here.. I am trying to work on a minimal reproduction of the problem.

The issue seemed to have started for us on version 1.10 of the SDK after we made a combination of changes:

1. Changed our DocumentDB collection to use Partitioning. At which point we received an error stating that our processes need to run as 64-bit.
2. Updating the Azure Website to run as 64-bit. It was previously configured as 32-bit. 

Are we able to query against partitioned collections in a 32-bit process with SDK version 1.11.3? If so, I could try upgrading to that version and then changing the Azure Website to 32-bit in an attempt to rule out some configuration.

This is currently a pretty big problem for us, as it is preventing us from being able to reliably deploy our applications.

**Comment by vijayrkn** [comment]
The appOffline issue is already fixed in Azure.

Is the app deployed to Azure or a different host? If it is not Azure, then you will have to install the latest ANCM - https://github.com/aspnet/Home/issues/694#issuecomment-276761626.

**Comment by SteMcLellan** [comment]
@vijayrkn I have tried to read through those issues, but I am not sure if they apply to me. My application is not running on ASP .NET Core. I've also tried all the work around in the mentioned github issues.

From what I can gather, the problem seems to happen when executing a cross partitioned query. I've been able to create a small ASP Web API project that has a single method that does a cross partitioned query. After executing the query I can log into the Kudu debug console, and attempting to re-save the `DocumentDB.Spatial.Sql.dll` issues an error saying the file is locked. From this point forward I cannot deploy back to the site and overwrite that file until I start/stop the app service.

**Comment by vijayrkn** [comment]
@SteMcLellan - Now I understand the issue. 

For the runtime issue causing the dll to get locked, please create an issue in the ASP.NET MVC/Web API repo (If this issue is with ASP.NET Web API).

For now to get past the deployment issue, you can set this property to the pubxml (if publishing from VS) or pass this property to the build. This will take care of bringing the site down before deployment (so that manually re-starting the site is not necessary). 

```<EnableMSDeployAppOffline>true</EnableMSDeployAppOffline>```

Hope this helps!

**Comment by rnagpal** [comment]
@SteMcLellan DocumentDB SDK release 1.11.3 supports cross partition queries in 32 bit mode, so you should be good. Give it a try do see if this is a 64-bit related issue. 

@vijayrkn AFAIRC, I did tried <EnableMSDeployAppOffline>true</EnableMSDeployAppOffline> when publishing from VS and I was still getting the lock issue. Is there a VS update that would fix that or is this an Azure issue that has been fixed recently?

**Comment by SteMcLellan** [comment]
@rnagpal I am not able to execute my partitioned query against my database in an application running in 32-bit mode, using SDK 1.13.3 as I receive an error. I receive the error both through an Azure Website and a Console application. I run the console app on my laptop. The error I receive is:

> Unhandled Exception: System.AggregateException: One or more errors occurred. ---> System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (Exception from HRESULT: 0x8007000B)
>    at Microsoft.Azure.Documents.ServiceInteropWrapper.CreateServiceProvider(String configJsonString, IntPtr& serviceProvider)
>    at Microsoft.Azure.Documents.Query.QueryPartitionProvider.Initialize()
>    at Microsoft.Azure.Documents.Query.QueryPartitionProvider.GetPartitionedQueryExecutionInfoInternal(SqlQuerySpec querySpec, PartitionKeyDefinition partitionKeyDefinition, Boolean requireFormattableOrderByQuery, Boolean isContinuationExpected)
>    at Microsoft.Azure.Documents.Routing.PartitionRoutingHelper.GetProvidedPartitionKeyRanges(SqlQuerySpec querySpec, Boolean enableCrossPartitionQuery, Boolean parallelizeCrossPartitionQuery, PartitionKeyDefinition partitionKeyDefinition, QueryPartitionProvider queryPartitionProvider, String clientApiVersion, QueryInfo& queryInfo)
>    at Microsoft.Azure.Documents.Query.DefaultDocumentQueryExecutionContext.<TryGetTargetPartitionKeyRangeAsync>d__18.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---

If I uncheck `Prefer 32-bit` checkbox in VS then the program will execute w/o error.

The app basically has a single method which executes the query:

```
            DocumentClient D = new DocumentClient(new Uri(DOCDB_URI), DOCDB_KEY);

            Uri documentCollectionUri = UriFactory.CreateDocumentCollectionUri(DOCDB_DB, DOCDB_COL);

            var result = D.CreateDocumentQuery<SillyDocument>(documentCollectionUri, new FeedOptions { EnableCrossPartitionQuery = true })
                            .Where(S => S.type == "company")
                            .AsDocumentQuery();

            while (result.HasMoreResults)
            {
                var x = await result.ExecuteNextAsync<SillyDocument>();
                Console.WriteLine("Found results");
                Console.ReadKey();
                return;
            }

            Console.WriteLine("Found nothing");
            Console.ReadKey();

```

Is there anything wrong with the above code? Are you able to replicate this?

**Comment by rnagpal** [comment]
@SteMcLellan You are right. We have a bug here and I'm working on a fix for that. Someone else also reported that today here: https://github.com/Azure/azure-documentdb-dotnet/issues/140 
In 1.11.3, we fixed running queries against single partition collections in 32 bit mode that was broken in 1.11.* releases. 
As a workaround, can you uncheck Prefer 32 bit option until then into your app.
We would be doing a release early next week to fix this issue.

**Comment by rnagpal** [comment]
@SteMcLellan We have released a fix for the 32-bit issue you reported as part of 1.11.4 release. Please upgrade and confirm. If you still see the file lock issue, feel free to create a new issue in the ASP.NET MVC/Web API repo as @vijayrkn recommended and just add a reference to that in this issue, so that people can track it, if they run into this issue.

**Comment by SteMcLellan** [comment]
@rnagpal Thanks!

I can confirm that I can now execute partitioned queries in 32-bit mode. Additionally, if I run my website in 32-bit mode, `DocumentDb.Spatial.Sql.dll` does not get locked. Thanks for your help.

**Comment by rnagpal** [comment]
@Jacob-McKay Let me know if you too are not seeing the file lock issue with the latest version.
